Bulk actions on UI

klakhov opened the pull request:
<!-- Raise an issue to propose your change (https://github.com/cvat-ai/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [Contribution guide](https://docs.cvat.ai/docs/contributing/). -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->
The PR introduces a way to select multiple resources and make a bulk operation on full group.

Depends on #9543, #9527
### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable for some reason, then ~~explicitly strikethrough~~ the whole
line. If you don't do that, GitHub will show incorrect progress for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment <!-- see top comment in CHANGELOG.md -->
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

bsekachev commented:
When I selected 1,2,3 with Ctrl and then holding shift I click number 5. 
It should select 1,2,3,4,5, not 3,4,5.

bsekachev commented:
BUG:  Any click here resets the selection and bulk export 

<img width="1334" height="808" alt="image" src="https://github.com/user-attachments/assets/a915dc4b-5496-46df-b0eb-42b19952d400" />

bsekachev commented:
Performance: Let's send preview requests one by one. Otherwise we are risky to ddos ourselves.

bsekachev commented:
Missing padding 
<img width="396" height="63" alt="image" src="https://github.com/user-attachments/assets/4d79fe58-cb5f-4a11-a23e-312d98430931" />

bsekachev commented on changelog.d/20250708_134458_klakhov_bulk_actions.md:
```diff
@@ -0,0 +1,5 @@
+### Added
+
+- Introduced bulk actions to perform single operations on multiple selected resources at once
```
```suggestion
- Introduced bulk actions to perform operations on multiple selected resources at once
```

bsekachev replied:
Let's specify more details about supported bulk operations

klakhov replied:
Added more info

bsekachev commented on cvat-ui/src/components/task-page/task-page.tsx:
```diff
@@ -37,19 +37,26 @@ function TaskPageComponent(): JSX.Element {
                 if (task) {
                     setTaskInstance(task);
+                    if (task.jobs && Array.isArray(task.jobs)) {
+                        const jobsList = [...task.jobs] as JobsList;
```
task.jobs already have specified type: `Job[]` and it is not compatible with JobsList having also count. So, this cast does not seem to be correct. 

Maybe to rewrite it just as:

```
dispatch(jobsActions.getJobsSuccess(
    Object.defineProperty([...task.jobs], 'count', { value: ... })
));
```

bsekachev commented on cvat-ui/src/utils/platform.ts:
```diff
@@ -0,0 +1,18 @@
+
+    if (window.isSecureContext && 'userAgentData' in navigator) {
+        // Use the new User-Agent Client Hints API when available in secure context
+        platformName = (navigator as any).userAgentData?.platform || navigator.platform;
```
With just ``if .. else`` it may be rewritten with two returns. 'unknown' is not necessary in the beginning.
Additionally, I believe .toLowerCase() should not be in utility function. If outside code needs lowercase, better to do it outside.

bsekachev commented on cvat-ui/src/index.tsx:
```diff
@@ -123,10 +129,13 @@ function mapDispatchToProps(dispatch: any): DispatchToProps {
         loadOrganization: (): void => dispatch(activateOrganizationAsync()),
-        initInvitations: (): void => dispatch(getInvitationsAsync({ page: 1 }, true)),
-        initRequests: (): void => dispatch(getRequestsAsync({ page: 1 })),
+        initInvitations: (): void => dispatch(getInvitationsAsync({ page: 1, pageSize: 10 }, true)),
+        initRequests: (): void => dispatch(getRequestsAsync({ page: 1, pageSize: 10 })),
```
Should be initialized from InvitationsState/RequestsState instead of hard-coding 10 here.

bsekachev replied:
BTW, `pageSize` is missing in default value for InvitationdState and typescript complaints about it

bsekachev commented on cvat-ui/src/components/cvat-app.tsx:
```diff
@@ -456,7 +471,40 @@ class CVATApplication extends React.PureComponent<CVATAppProps & RouteComponentP
             }
         }
-
+        const bulkError = notifications.errors.selection.bulkOperation;
```
I believe bulk-progress component should be responsible for this logic. Like:

... progress ...
... N items failed, repeat? ... 

What do you think?

bsekachev commented on cvat-ui/src/components/bulk-progress.tsx:
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: MIT
+
+import React from 'react';
+import { Progress, Button } from 'antd';
```
Maybe this is not necessary anymore (need to double check), but let's stick to current style of importing each component from `antd` separately.

bsekachev commented on cvat-ui/src/components/bulk-progress.tsx:
```diff
@@ -0,0 +1,36 @@
+
+export default function BulkProgress(): JSX.Element | null {
+    const dispatch = useDispatch();
+    const fetching = useSelector((state: CombinedState) => state.selection.fetching);
+    const status = useSelector((state: CombinedState) => state.selection.status);
```
Please, combine useSelectors.

bsekachev commented on cvat-ui/src/components/bulk-progress.tsx:
```diff
@@ -0,0 +1,36 @@
+            </div>
+            <Button
+                className='cvat-bulk-progress-cancel'
+                danger
```
I do not think it should be `danger` by concept.

bsekachev commented on cvat-ui/src/reducers/index.ts:
```diff
@@ -257,6 +258,18 @@ export interface CloudStoragesState {
     };
 }
 
+export interface SelectionStatus {
```
It seems this is not SelectionStatus, this is BulkActionStatus

bsekachev commented on cvat-ui/src/reducers/index.ts:
```diff
@@ -257,6 +258,18 @@ export interface CloudStoragesState {
+export interface SelectionState {
+    selected: (number | string)[];
+    fetching: boolean;
+    status: { message: string; percent: number } | null;
```
SelectionStatus (BulkActionStatus) | null ?

bsekachev commented on cvat-ui/src/reducers/index.ts:
```diff
@@ -645,6 +658,17 @@ export interface NotificationsState {
+        };
+        selection: {
+            bulkOperation: {
+                message: string,
+                remainingItemsCount: number,
```
```suggestion
                message: string;
                remainingItemsCount: number;
```

etc...

bsekachev commented on cvat-ui/src/reducers/index.ts:
```diff
@@ -1000,6 +1024,9 @@ export interface OrganizationState {
     leaving: boolean;
     removingMember: boolean;
     updatingMember: boolean;
+    gettingMembers: boolean;
```
```suggestion
    fetchingMembers: boolean;
```

bsekachev commented on cvat-ui/src/reducers/index.ts:
```diff
@@ -645,6 +658,17 @@ export interface NotificationsState {
             canceling: null | ErrorState;
             deleting: null | ErrorState;
+        };
+        selection: {
+            bulkOperation: {
```
I would propose to split `selection` logic and `bulkOperation` logic on state level. 
E.g. now in this state nothing is responsible for `selection`, it is only about bulk operation.

bsekachev replied:
Selection is only responsible for adding/removing elements from selection IMO. 
Bulk operation is responsible for progress/status/fails.

klakhov replied:
Makes sence, removed.

bsekachev commented on cvat-ui/src/reducers/selection-reducer.ts:
```diff
@@ -0,0 +1,52 @@
+                return { ...state, selected: Array.from(new Set([...state.selected, ...action.payload.resourceIds])) };
+            }
+            return { ...state, selected: [...action.payload.resourceIds] };
+        case SelectionActionsTypes.CLEAR_SELECTED_RESOURCES:
+            return { ...state, selected: [] };
```
Seems like redundant action. You may just call SELECT_RESOURCES with empty array.

klakhov replied:
I’d prefer to keep this action. The current `Select` action works as **“append some resources”**, not **“fully replace selected resources.”**

This is useful when we want to append items (e.g., via Ctrl or Shift click), so we don’t need to worry about some previously selected ones. It keeps the "set" logic centralized at the reducer level.

bsekachev commented on cvat-ui/src/reducers/selection-reducer.ts:
```diff
@@ -0,0 +1,52 @@
+    cancelled: false,
+};
+
+export default function selectionReducer(
```
As mentioned before, each state (Projects/Tasks ... etc) should store selected items. 
In this case this one would be `bulkOperationReducer`. 

Select/Deselect may be a separated ActionTypes supported by different reducers in common code.

bsekachev commented on cvat-ui/src/styles.scss:
```diff
@@ -210,4 +210,49 @@ html, body {
+}
+
+.cvat-bulk-progress-message {
+    margin-top: 24px;
```
Please, use grid-unit-size in styles

klakhov commented:
- Fixed shift behaviour, now it only appends
- Fixed arrow buttons for dynamic selector and now its hidden right after selection
- Optimized preview fetching in #9692
- "Reset assignee" was shown only in 'bulk mode'(when several elements were selected), changed it now it is always shown
- Added padding in info icon
- Now when request is cancelled we will show 'Cancelled' status 
- Removed selection of cancelled/deleted resources with 'select all'
